File: packages/guards/src/paragraph-structure.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ParagraphStructureOptions {
  action: 'block' | 'warn';
  maxWords?: number;
}

function countWords(s: string): number {
  return s.trim().split(/\s+/).filter((w) => w.length > 0).length;
}

export function paragraphStructure(options: ParagraphStructureOptions): Guard {
  const maxWords = options.maxWords ?? 200;

  return {
    name: 'paragraph-structure',
    version: '0.1.0',
    description: 'Validates paragraph structure',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];
      const paragraphs = text.split(/\n\s*\n/).filter((p) => p.trim().length > 0);

      for (let i = 0; i < paragraphs.length; i++) {
        const p = paragraphs[i].trim();
        if (/^#{1,6}\s/.test(p) || /^[-*+]\s/.test(p) || /^\|/.test(p)) continue;

        const wc = countWords(p);
        if (wc > maxWords) {
          issues.push(`Paragraph ${i + 1} too long (${wc} words, max ${maxWords})`);
        }

        const sentences = p.split(/(?<=[.!?])\s+/).filter((s) => s.trim().length > 0);
        if (sentences.length === 1 && wc > 10) {
          issues.push(`Paragraph ${i + 1} has only one sentence`);
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'paragraph-structure',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, paragraphCount: paragraphs.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/paragraph_structure.py
"""Validate paragraph structure."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult


def _count_words(s: str) -> int:
    return len(
        [w for w in s.strip().split() if w]
    )


class _ParagraphStructure:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_words: int = 200,
    ) -> None:
        self.name = "paragraph-structure"
        self.action = action
        self._max_words = max_words

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []
        paragraphs = [
            p.strip()
            for p in re.split(r"\n\s*\n", text)
            if p.strip()
        ]

        for i, p in enumerate(paragraphs):
            if re.match(r"^#{1,6}\s", p):
                continue
            if re.match(r"^[-*+]\s", p):
                continue
            if p.startswith("|"):
                continue

            wc = _count_words(p)
            if wc > self._max_words:
                issues.append(
                    f"Paragraph {i+1} too long"
                    f" ({wc} words,"
                    f" max {self._max_words})"
                )

            sents = [
                s.strip()
                for s in re.split(
                    r"(?<=[.!?])\s+", p
                )
                if s.strip()
            ]
            if len(sents) == 1 and wc > 10:
                issues.append(
                    f"Paragraph {i+1} has"
                    " only one sentence"
                )

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="paragraph-structure",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Paragraph structure issues found"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "paragraph_count": len(paragraphs),
                }
                if triggered
                else None
            ),
        )


def paragraph_structure(
    *,
    action: str = "warn",
    max_words: int = 200,
) -> _ParagraphStructure:
    return _ParagraphStructure(
        action=action, max_words=max_words
    )
